include/roboteam_utils/HalfLine.h

```diff
@@ -0,0 +1,36 @@
+#ifndef RTT_ROBOTEAM_UTILS_SRC_HALFLINE_H_
+#define RTT_ROBOTEAM_UTILS_SRC_HALFLINE_H_
+
+#include "Line.h"
+#include "Vector2.h"
+
+namespace rtt {
+/**
+ * The HalfLine class represents a line that is semi-infinite long, i.e. it has an end point where the line stops and an direction in which it is infinite long.
+ *
+ * @author Haico Dorenbos
+ * @since 2020-04-22
+ */
+class HalfLine {
+    private:
+        const Vector2 ending; // The location where the HalfLine ends
+        const Vector2 direction; // This vector indicates in which direction the HalfLine is infinite long
```

Comment:
I don't really see why these should be constant. If we want to be consistent with other geometry classes these would be mutable.

Reply:
Well geometry classes in my point of view should always be immutable

Reply:
Then that also means we would e.g. have that vector2+=Vector2 is not valid, which in my opinion would be a huge loss, as it means we cannot do inplace affine transformations anymore anywhere. 
We want users to be able to freely move around objects without them having to create new ones. This means functions that do alter the object are not marked as constant, and other functions are. This will also cause unnecessary performance loss when you want to for example generate/move around obstacles as part of path planning as you need to call the constructor again. Having a shape have the function += Vector2 to apply an affine transformation on it is massively beneficial to have.
It also prevents usage such as :

Vector x;
if(predicate){
...
x = someCalculation()
}else {
...
x = someOtherCalculation()
}

So I personally don't really think this is a good idea to make all geometry classes immutable. It will in the end cause in my opinion more ugly code, especially if on the places of (...) you also need to do other calculations. (Since then the inline if statement becomes problematic.)

Reply:
I guess you are both not familiar with the concept of being an immutable class. The thing about immutable classes is that variables inside the class cannot be changed/reassigned. Variables outside the class can be changed/reassigned however. The example you gave:

Vector x;
if(predicate){
...
x = someCalculation()
}else {
...
x = someOtherCalculation()
}

is still possible if Vector is an immutable class, because variable x is mutable and if someCalculation() passes a new instance of Vector then it is just fine. Also currently the +=, -= operation pass a new instance of Vector (but also update the old version which I consider to be a defect), so I do not see what the problem is. Last of all I don't really see the reason why creating new class instances is bad, that is just part of programming that you constantly create new class instances. On the other hand having these class mutable is dangerous, since if you for example have:

Vector x = something();
Vector y = somethingElse();
y += x;

Then developers assume that x still got the same value, but it does not since += is implemented to change x as well.

Reply:
I discussed this in detail with Rolf. There is currently a bug in the +=, *=, -=, etc. operators in most classes since they create a new instance rather than returning a reference to the updated old one. I agree with Rolf that geometry classes might be mutable (for the time being).